File: ts/project/.sdk/tm/lua/feature/telemetry_feature.lua
-- ProjectName SDK telemetry feature
--
-- Distributed-tracing telemetry. Opens a span per operation (PrePoint),
-- propagates trace context to the server as W3C `traceparent` plus
-- `X-Trace-Id` / `X-Span-Id` headers (PreRequest), and closes the span on
-- completion (PreDone) or failure (PreUnexpected) — exactly once per
-- operation. Finished spans are kept on `client._telemetry.spans`; an
-- `exporter` callback, when provided, is invoked with each finished span.
-- Trace/span id generation (`idgen`) and the clock (`now`) are injectable
-- for deterministic tests.
--
-- Note: `end` is a Lua keyword, so the span close time is stored under the
-- string key `span["end"]`.

local BaseFeature = require("feature.base_feature")

local TelemetryFeature = {}
TelemetryFeature.__index = TelemetryFeature
setmetatable(TelemetryFeature, { __index = BaseFeature })


function TelemetryFeature.new()
  local self = setmetatable(BaseFeature.new(), TelemetryFeature)
  self.version = "0.0.1"
  self.name = "telemetry"
  self.active = true
  self.client = nil
  self.options = nil
  -- Weak keys: an abandoned context must not leak its open span.
  self.spans = setmetatable({}, { __mode = "k" })
  self.seq = 0
  return self
end


function TelemetryFeature:init(ctx, options)
  self.client = ctx.client
  self.options = options or {}

  if options["active"] == true then
    self.active = true
  else
    self.active = false
  end

  self.spans = setmetatable({}, { __mode = "k" })
  self.seq = 0

  if not self.active then
    return
  end

  local client = self.client
  if client._telemetry == nil then
    client._telemetry = { spans = {}, active = 0 }
  end
end


function TelemetryFeature:PrePoint(ctx)
  if not self.active then
    return
  end
  local span = {
    traceId = self:_id("trace"),
    spanId = self:_id("span"),
    name = ((ctx.op ~= nil and ctx.op.entity) or "_") .. "." ..
      ((ctx.op ~= nil and ctx.op.name) or "_"),
    start = self:_now(),
  }
  self.spans[ctx] = span
  local client = self.client
  client._telemetry.active = client._telemetry.active + 1
end


function TelemetryFeature:PreRequest(ctx)
  if not self.active then
    return
  end
  local span = self.spans[ctx]
  local spec = ctx.spec
  if span == nil or spec == nil then
    return
  end
  if spec.headers == nil then
    spec.headers = {}
  end
  local h = self.options["headers"] or {}
  spec.headers[h["trace"] or "X-Trace-Id"] = span.traceId
  spec.headers[h["span"] or "X-Span-Id"] = span.spanId
  spec.headers[h["parent"] or "traceparent"] =
    "00-" .. span.traceId .. "-" .. span.spanId .. "-01"
end


function TelemetryFeature:PreDone(ctx)
  if not self.active then
    return
  end
  local ok = ctx.result ~= nil and ctx.result.ok ~= false and ctx.result.err == nil
  self:_close(ctx, ok)
end


function TelemetryFeature:PreUnexpected(ctx)
  if not self.active then
    return
  end
  self:_close(ctx, false)
end


function TelemetryFeature:_close(ctx, ok)
  -- Close once per operation; a PreDone followed by a pipeline error
  -- (non-2xx) fires PreUnexpected too, which then finds no open span.
  local span = self.spans[ctx]
  if span == nil then
    return
  end
  self.spans[ctx] = nil
  span["end"] = self:_now()
  span.durationMs = math.max(0, span["end"] - span.start)
  span.ok = ok

  local client = self.client
  client._telemetry.active = client._telemetry.active - 1
  table.insert(client._telemetry.spans, span)

  local exporter = self.options["exporter"]
  if type(exporter) == "function" then
    pcall(exporter, span)
  end
end


function TelemetryFeature:_id(kind)
  local idgen = self.options["idgen"]
  if type(idgen) == "function" then
    return idgen(kind)
  end
  -- Deterministic-ish sequential id; unique within a client instance.
  self.seq = self.seq + 1
  local n = string.format("%04x", self.seq)
  local prefix = kind == "trace" and "t" or "s"
  return prefix .. n .. string.rep("0", 12)
end


function TelemetryFeature:_now()
  local now = self.options["now"]
  if type(now) == "function" then
    return now()
  end
  return os.time() * 1000
end


return TelemetryFeature

File: ts/project/.sdk/tm/php/feature/TelemetryFeature.php
<?php
declare(strict_types=1);

// ProjectName SDK telemetry feature

require_once __DIR__ . '/BaseFeature.php';

// Distributed-tracing telemetry. Opens a span per operation (PrePoint),
// propagates trace context to the server as W3C `traceparent` plus
// `X-Trace-Id` / `X-Span-Id` headers (PreRequest), and closes the span on
// completion (PreDone) or failure (PreUnexpected) — once per operation.
// Finished spans are kept on `client->_telemetry['spans']`; an `exporter`
// callback, when provided, is invoked with each finished span. Trace/span
// id generation (`idgen`) and the clock (`now`) are injectable for
// deterministic tests. Mirrors ts/src/feature/telemetry/TelemetryFeature.ts.
class ProjectNameTelemetryFeature extends ProjectNameBaseFeature
{
    private mixed $client;
    private ?array $options;
    private ?\WeakMap $spans;
    private int $seq;

    public function __construct()
    {
        parent::__construct();
        $this->version = '0.0.1';
        $this->name = 'telemetry';
        $this->active = true;
        $this->client = null;
        $this->options = null;
        $this->spans = null;
        $this->seq = 0;
    }

    public function init(ProjectNameContext $ctx, array $options): void
    {
        $this->client = $ctx->client;
        $this->options = $options;
        $this->active = ($options['active'] ?? null) === true;
        $this->spans = new \WeakMap();
        $this->seq = 0;

        if (!$this->active) {
            return;
        }

        $client = $this->client;
        if (!isset($client->_telemetry)) {
            $client->_telemetry = ['spans' => [], 'active' => 0];
        }
    }

    public function PrePoint(ProjectNameContext $ctx): void
    {
        if (!$this->active || $this->spans === null) {
            return;
        }
        $span = [
            'traceId' => $this->_id('trace'),
            'spanId' => $this->_id('span'),
            'name' => ($ctx->op->entity !== '' ? $ctx->op->entity : '_')
                . '.' . ($ctx->op->name !== '' ? $ctx->op->name : '_'),
            'start' => $this->_now(),
            'end' => null,
            'durationMs' => null,
            'ok' => null,
        ];
        $this->spans[$ctx] = $span;
        $this->client->_telemetry['active']++;
    }

    public function PreRequest(ProjectNameContext $ctx): void
    {
        if (!$this->active || $this->spans === null || !isset($this->spans[$ctx])) {
            return;
        }
        $span = $this->spans[$ctx];
        $spec = $ctx->spec;
        if ($spec === null) {
            return;
        }
        $h = $this->options['headers'] ?? [];
        if (!is_array($h)) {
            $h = [];
        }
        $spec->headers[$h['trace'] ?? 'X-Trace-Id'] = $span['traceId'];
        $spec->headers[$h['span'] ?? 'X-Span-Id'] = $span['spanId'];
        $spec->headers[$h['parent'] ?? 'traceparent'] =
            '00-' . $span['traceId'] . '-' . $span['spanId'] . '-01';
    }

    public function PreDone(ProjectNameContext $ctx): void
    {
        if (!$this->active) {
            return;
        }
        $ok = $ctx->result !== null && $ctx->result->ok !== false && $ctx->result->err === null;
        $this->_close($ctx, $ok);
    }

    public function PreUnexpected(ProjectNameContext $ctx): void
    {
        if (!$this->active) {
            return;
        }
        $this->_close($ctx, false);
    }

    private function _close(ProjectNameContext $ctx, bool $ok): void
    {
        // Close once per operation; a PreDone followed by a pipeline failure
        // fires PreUnexpected too, which then finds no open span.
        if ($this->spans === null || !isset($this->spans[$ctx])) {
            return;
        }
        $span = $this->spans[$ctx];
        unset($this->spans[$ctx]);
        $span['end'] = $this->_now();
        $span['durationMs'] = max(0.0, (float)$span['end'] - (float)$span['start']);
        $span['ok'] = $ok;

        $client = $this->client;
        $client->_telemetry['active']--;
        $client->_telemetry['spans'][] = $span;

        $exporter = $this->options['exporter'] ?? null;
        if (is_callable($exporter)) {
            try {
                $exporter($span);
            } catch (\Throwable $_e) {
                // Exporter failures must never break the pipeline.
            }
        }
    }

    private function _id(string $kind): string
    {
        $idgen = $this->options['idgen'] ?? null;
        if (is_callable($idgen)) {
            return (string)$idgen($kind);
        }
        // Deterministic-ish sequential id; unique within a client instance.
        $this->seq++;
        $n = str_pad(dechex($this->seq), 4, '0', STR_PAD_LEFT);
        return ($kind === 'trace' ? 't' : 's') . str_pad($n, 16, '0', STR_PAD_RIGHT);
    }

    private function _now(): float
    {
        $now = $this->options['now'] ?? null;
        if (is_callable($now)) {
            return (float)$now();
        }
        return microtime(true) * 1000.0;
    }
}
